File: code/scenes/src/main/java/org/fbme/scenes/controllers/diagram/ConnectionEntry.java
package org.fbme.scenes.controllers.diagram;

import org.fbme.scenes.controllers.SelectionModel;
import org.jetbrains.annotations.Nullable;

import java.awt.*;
import java.util.Set;
import java.util.function.BiFunction;

class ConnectionEntry<CompT, PortT, ConnT, CursorT, PathT> {
    private final ConnectionsFacility<CompT, PortT, ConnT, CursorT, PathT> myConnFacility;
    final ConnT myConnection;

    final ConnectionController<CursorT, PathT> myController;

    PathT myModelPath;
    PathT myTransformedPath;

    private final BiFunction<Point, Point, PathT> myPathProvider;

    ConnectionEntry(ConnectionsFacility<CompT, PortT, ConnT, CursorT, PathT> connFacility, ConnT connection) {
        myConnFacility = connFacility;
        myConnection = connection;
        myController = myConnFacility.myControllerFactory.create(myConnFacility.myScene.getContext(), myConnection);
        myPathProvider = myConnFacility.myConnectionSynchronizer.getPath(myConnection);
        DiagramController<CompT, PortT, ConnT> diagramController = myConnFacility.myDiagramController;
        PortController sourcePort = diagramController.getPortController(diagramController.getSource(myConnection));
        PortController targetPort = diagramController.getPortController(diagramController.getTarget(myConnection));
        myModelPath = myPathProvider.apply(sourcePort.getModelEndpointPosition(), targetPort.getModelEndpointPosition());
    }


    void relayout() {
        DiagramController<CompT, PortT, ConnT> diagramController = myConnFacility.myDiagramController;
        PortController sourcePort = diagramController.getPortController(diagramController.getSource(myConnection));
        PortController targetPort = diagramController.getPortController(diagramController.getTarget(myConnection));
        myModelPath = myPathProvider.apply(sourcePort.getModelEndpointPosition(), targetPort.getModelEndpointPosition());
        myController.updateCellWithPath((myTransformedPath != null ? myTransformedPath : myModelPath));
        myController.updateCellSelection(isSelected());
    }

    Rectangle getLayoutBounds() {
        Rectangle bounds = myController.getBounds(myModelPath);
        if (myTransformedPath != null) {
            bounds = bounds.union(myController.getBounds(myTransformedPath));
        }
        return bounds;
    }

    @Nullable
    CursorT getCursor() {
        if (myConnFacility.myConnectionCursor == null || myConnFacility.myConnectionCursor.getConnection() != myConnection) {
            return null;
        }
        return myConnFacility.myConnectionCursor.getCursor();
    }

    boolean isSelected() {
        Set<ConnT> myConnectionsSelection = myConnFacility.myConnectionsSelection;
        SelectionModel<CompT> componentsSelection = myConnFacility.myComponentsSelection;
        DiagramController<CompT, PortT, ConnT> diagramController = myConnFacility.myDiagramController;
        return myConnectionsSelection.contains(myConnection) ||
                componentsSelection.isSelected(diagramController.getComponent(diagramController.getSource(myConnection))) ||
                componentsSelection.isSelected(diagramController.getComponent(diagramController.getTarget(myConnection)));
    }
}

File: code/scenes/src/main/kotlin/org/fbme/scenes/controllers/diagram/ConnectionEntry.kt
package org.fbme.scenes.controllers.diagram

import java.awt.Point
import java.awt.Rectangle

internal class ConnectionEntry<CompT, PortT, ConnT, CursorT, PathT>(
    private val connectionsFacility: ConnectionsFacility<CompT, PortT, ConnT, CursorT, PathT>,
    val connection: ConnT
) {
    val controller: ConnectionController<CursorT, PathT> =
        connectionsFacility.controllerFactory.create(connectionsFacility.scene.editorContext, connection)
    private val pathProvider: (Point, Point) -> PathT = connectionsFacility.connectionSynchronizer.getPath(connection)
    var modelPath: PathT
    var transformedPath: PathT? = null

    fun relayout() {
        val diagramController = connectionsFacility.diagramController
        val sourcePort = diagramController.getSource(connection) ?: error("Source not found")
        val targetPort = diagramController.getTarget(connection) ?: error("Target not found")
        val sourcePortController = diagramController.getPortController(sourcePort)
        val targetPortController = diagramController.getPortController(targetPort)
        modelPath = pathProvider(sourcePortController.modelEndpointPosition, targetPortController.modelEndpointPosition)
        controller.updateCellWithPath(transformedPath ?: modelPath)
        controller.updateCellSelection(isSelected)
    }

    val layoutBounds: Rectangle
        get() {
            val bounds = controller.getBounds(modelPath)
            return if (transformedPath == null) bounds else bounds.union(controller.getBounds(transformedPath!!))
        }
    val cursor: CursorT?
        get() {
            val connectionCursor = connectionsFacility.connectionCursor
            if (connectionCursor != null && connectionCursor.connection == connection) {
                return connectionCursor.cursor
            }
            return null
        }
    val isSelected: Boolean
        get() {
            val myConnectionsSelection = connectionsFacility.connectionsSelection
            val componentsSelection = connectionsFacility.componentsSelection
            val diagramController = connectionsFacility.diagramController
            val sourcePort = diagramController.getSource(connection)
            val targetPort = diagramController.getTarget(connection)
            val isSourceComponentSelected = sourcePort?.let {
                componentsSelection.isSelected(diagramController.getComponent(it))
            } ?: false
            val isTargetComponentSelected = targetPort?.let {
                componentsSelection.isSelected(diagramController.getComponent(it))
            } ?: false
            return myConnectionsSelection.contains(connection) ||
                isSourceComponentSelected ||
                isTargetComponentSelected
        }

    init {
        val diagramController = connectionsFacility.diagramController
        val sourcePort = diagramController.getSource(connection) ?: error("Source not found")
        val targetPort = diagramController.getTarget(connection) ?: error("Target not found")
        val sourcePortController = diagramController.getPortController(sourcePort)
        val targetPortController = diagramController.getPortController(targetPort)
        modelPath = pathProvider(sourcePortController.modelEndpointPosition, targetPortController.modelEndpointPosition)
    }
}
